src/conn.rs

```diff
@@ -366,14 +379,20 @@ impl<'a, 'c> InProgress<'a, 'c> {
     }
 
     pub fn transact_terms<I>(&mut self, terms: I, tempid_set: InternSet<TempId>) -> Result<TxReport> where I: IntoIterator<Item=TermWithTempIds> {
-        let (report, next_partition_map, next_schema, _watcher) =
+        let w = InProgressTransactWatcher::new(
+                InProgressObserverTransactWatcher::new(self.tx_observer),
+                self.cache.transact_watcher());
+        let (report, next_partition_map, next_schema, mut watcher) =
             transact_terms(&self.transaction,
                            self.partition_map.clone(),
                            &self.schema,
                            &self.schema,
-                           self.cache.transact_watcher(),
+                           w,
                            terms,
                            tempid_set)?;
+        if let Some(tx_id) = watcher.tx_id() {
```

Comment:
Further to comment about `tx_id`, I think what you mean here is something like `w.observer_watcher.transacted_tx()`… but I think _actually_ what you want is to _reuse_ the watcher across the lifetime of the `InProgress`, and the watcher collects the `tx_ids`, no?